addons/sale_stock/models/account_move.py

```diff
@@ -25,86 +26,59 @@ def _stock_account_get_last_step_stock_moves(self):
-
-        # Prepare and return lot_values
-        qties_per_lot = defaultdict(lambda: 0)
-        if self.type == 'out_refund':
-            for ml in outgoing_sml:
-                qties_per_lot[ml.lot_id] -= ml.product_uom_id._compute_quantity(ml.qty_done, ml.product_id.uom_id)
-            for ml in incoming_sml:
-                qties_per_lot[ml.lot_id] += ml.product_uom_id._compute_quantity(ml.qty_done, ml.product_id.uom_id)
-        else:
-            for ml in outgoing_sml:
-                qties_per_lot[ml.lot_id] += ml.product_uom_id._compute_quantity(ml.qty_done, ml.product_id.uom_id)
-            for ml in incoming_sml:
-                qties_per_lot[ml.lot_id] -= ml.product_uom_id._compute_quantity(ml.qty_done, ml.product_id.uom_id)
+        amls = self.invoice_line_ids.filtered(lambda aml: not aml.display_type and aml.product_id and aml.quantity)
+        all_amls = amls.sale_line_ids.invoice_lines.filtered(lambda aml: aml.move_id.state not in ('draft', 'cancel'))
+        # Use the default order of the AMLs to get the first AML of the current invoice
+        # and deduce all the previous ones
+        index = all_amls.ids.index(amls[:1].id) if amls[:1] in all_amls else 0
+        previous_amls = all_amls[:index]
+
+        previous_qties_invoiced = previous_amls._get_invoiced_qty_per_product()
+        invoiced_qties = amls._get_invoiced_qty_per_product()
+        invoiced_qties_copy = {**invoiced_qties}
+
+        qties_per_lot = defaultdict(float)
+        smls = amls.sale_line_ids.move_ids.move_line_ids.filtered(lambda sml: sml.state == 'done' and sml.lot_id).sorted(lambda sml: (sml.date, sml.id))
+        for sml in smls.filtered(lambda sml: sml.product_id in invoiced_qties.keys()):
+            product_uom = sml.product_id.uom_id
+            qty_done = sml.product_uom_id._compute_quantity(sml.qty_done, product_uom)
+
+            previous_qty_invoiced = previous_qties_invoiced.get(sml.product_id, 0)
+            if float_compare(previous_qty_invoiced, 0, precision_rounding=product_uom.rounding) > 0:
+                if float_compare(qty_done, previous_qty_invoiced, precision_rounding=product_uom.rounding) > 0:
+                    qties_per_lot[sml.lot_id] += qty_done - previous_qty_invoiced
+
+                if sml.location_id.usage == 'customer':
+                    previous_qties_invoiced[sml.product_id] += qty_done
+                elif sml.location_dest_id.usage == 'customer':
+                    previous_qties_invoiced[sml.product_id] -= qty_done
+
+            elif float_compare(invoiced_qties[sml.product_id], 0, precision_rounding=product_uom.rounding) > 0:
+                if sml.location_id.usage == 'customer':
+                    invoiced_qties[sml.product_id] += qty_done
+                elif sml.location_dest_id.usage == 'customer':
+                    invoiced_qties[sml.product_id] -= qty_done
+                qties_per_lot[sml.lot_id] += qty_done
+
```

Comment:
the inner if/elif seems a bit repetitive.
If you can extract it, it would be nice but that's not very important.

Reply:
Sorry but that's not what I meant by extracting the logic. :/

I was more thinking about conditional assign the dict to a variable and do the process after that. 
(And I'm not sure it's possible to simply do it)

In the current state, it's more complicated.

Reply:
Should be ok now